subsystem/drivetrain.py

```diff
@@ -0,0 +1,165 @@
+            math.radians(self.encoder.getAbsolutePosition())
+            - self.absolute_encoder_zeroed_pos
+        )
+
+        self.m_turn.set_sensor_position(
+            current_pos_rad * constants.drivetrain_turn_gear_ratio / (2 * math.pi)
+        )
+
+        self.m_move.set_sensor_position(0)
+        self.m_move.set_target_position(0)
+
+    def zero(self):
+        current_angle = self.get_current_motor_angle()
+
+        current_pos_rad = (
+            math.radians(self.encoder.getAbsolutePosition())
+            - self.absolute_encoder_zeroed_pos
+        )
+
+        self.m_turn.set_sensor_position(
+            current_pos_rad * constants.drivetrain_turn_gear_ratio / (2 * math.pi)
+        )
+
+        self.set_motor_angle(current_angle)
+
+    def raw_output(self, power):
+        self.m_move.set_raw_output(power)
+
+    def set_motor_angle(self, pos: radians):
+        self.m_turn.set_target_position(
+            (pos / (2 * math.pi)) * constants.drivetrain_turn_gear_ratio
+        )
+
+    def get_current_motor_angle(self) -> radians:
+        return (
+            (self.m_turn.get_sensor_position() / constants.drivetrain_turn_gear_ratio)
+            * 2
+            * math.pi
+        )
+
+    def set_motor_velocity(self, vel: meters_per_second):
+        self.m_move.set_target_velocity(vel * constants.drivetrain_move_gear_ratio)
+
+    def get_motor_velocity(self) -> radians_per_second:
+        return (
+            self.m_move.get_sensor_velocity()
+            / constants.drivetrain_move_gear_ratio_as_rotations_per_meter
+        )
+
+    def get_drive_motor_traveled_distance(self) -> meters:
+        sensor_position = -1 * self.m_move.get_sensor_position()
+
+        return (
+            sensor_position
+            / constants.drivetrain_move_gear_ratio_as_rotations_per_meter
+        )
+
+    def get_turn_motor_angle(self) -> radians:
+        return (
+            (self.m_turn.get_sensor_position() / constants.drivetrain_turn_gear_ratio)
+            * 2
+            * math.pi
+        )
+
+
+class Drivetrain(SwerveDrivetrain):
+    n_front_left = SparkMaxSwerveNode(
+        SparkMax(16, config=MOVE_CONFIG),
+        SparkMax(15, config=TURN_CONFIG),
+        CANCoder(24),
+        absolute_encoder_zeroed_pos=math.radians(174.638),
+        name="n_front_left",
+    )
+    n_front_right = SparkMaxSwerveNode(
+        SparkMax(14, config=MOVE_CONFIG),
+        SparkMax(13, config=TURN_CONFIG),
+        CANCoder(23),
+        absolute_encoder_zeroed_pos=math.radians(282.304),
+        name="n_front_right",
+    )
+    n_back_left = SparkMaxSwerveNode(
+        SparkMax(3, config=MOVE_CONFIG),
+        SparkMax(4, config=TURN_CONFIG),
+        CANCoder(21),
+        absolute_encoder_zeroed_pos=math.radians(313.769),
+        name="n_back_left",
+    )
+    n_back_right = SparkMaxSwerveNode(
+        SparkMax(5, config=MOVE_CONFIG),
+        SparkMax(6, config=TURN_CONFIG),
+        CANCoder(22),
```

Comment:
it is probably best to move the cad ids to the config so we can change them quickly